pkg/proxy/iptables/proxier.go

```diff
@@ -1060,20 +1060,23 @@ func (proxier *Proxier) syncProxyRules() {
 			}
 			activeNATChains[endpointChain] = true
 
-			args = append(args[:0], "-A", string(endpointChain))
-			args = proxier.appendServiceCommentLocked(args, svcNameString)
 			// Handle traffic that loops back to the originator with SNAT.
 			proxier.natRules.Write(
-				args,
+				"-A", string(endpointChain),
+				proxier.commentIfBelowThreshold(svcNameString),
 				"-s", epInfo.IP(),
 				"-j", string(KubeMarkMasqChain))
+			args = args[:0]
 			// Update client-affinity lists.
 			if svcInfo.SessionAffinityType() == v1.ServiceAffinityClientIP {
 				args = append(args, "-m", "recent", "--name", string(endpointChain), "--set")
 			}
 			// DNAT to final destination.
 			args = append(args, "-m", protocol, "-p", protocol, "-j", "DNAT", "--to-destination", epInfo.Endpoint)
-			proxier.natRules.Write(args)
+			proxier.natRules.Write(
+				"-A", string(endpointChain),
```

Comment:
I tried in some other places not to repeat the "-A <chain" when not needed, for fear of making a mistake.  It's not SUCH a big deal I guess.  This could also be:

```
args = append(args[:0], "-A", string(endpointChain))		
args = append(proxier.commentIfBelowThreshold(svcNameString)...)
```

I'm fine doing it your way if you think the win is worthwhile.  the point of the comment was to make `iptables-save | grep` useful.  In an earlier comment I suggested maybe comment-only rules.  That could help maybe?

```
-A KUBE-NOOP -m comment --comment "this chain does nothing" -j RETURN

-A KUBE-SEP-5QBLSAUHS7VAMKHU -m comment --comment "kube-system/metrics-server" -j KUBE-NOOP
-A KUBE-SEP-5QBLSAUHS7VAMKHU -s 10.64.3.3/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-5QBLSAUHS7VAMKHU -p tcp -m comment -m tcp -j DNAT ...
-A KUBE-SEP-5QBLSAUHS7VAMKHU -m comment --comment "kube-system/kube-dns:metrics" -j KUBE-NOOP
-A KUBE-SEP-6IIMKUPVCRULMU5Y -s 10.64.1.9/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-6IIMKUPVCRULMU5Y -p tcp -m tcp -j DNAT ...
-A KUBE-SEP-6QKF7KDE3IMNRERG -m comment --comment "default/hostnames" -j KUBE-NOOP
-A KUBE-SEP-6QKF7KDE3IMNRERG -s 10.64.2.7/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-6QKF7KDE3IMNRERG -p tcp -m tcp -j DNAT ...
-A KUBE-SEP-7X3ZD6IAFJB7UYBN -m comment --comment "kube-system/kube-dns:dns" -j KUBE-NOOP
-A KUBE-SEP-7X3ZD6IAFJB7UYBN -s 10.64.2.3/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-7X3ZD6IAFJB7UYBN -p udp -m udp -j DNAT ...
-A KUBE-SEP-CAIWW7T7PX3SGADF -m comment --comment "default/kubernetes:https" -j KUBE-NOOP
-A KUBE-SEP-CAIWW7T7PX3SGADF -s 104.155.184.204/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-CAIWW7T7PX3SGADF -p tcp -m tcp -j DNAT ...
-A KUBE-SEP-CPTDVYQW6QHZEJTA -m comment --comment "kube-system/kube-dns:metrics" -j KUBE-NOOP
-A KUBE-SEP-CPTDVYQW6QHZEJTA -s 10.64.2.3/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-CPTDVYQW6QHZEJTA -p tcp -m tcp -j DNAT ...
-A KUBE-SEP-DERWY4MTP2RTOITK -m comment --comment "kube-system/kube-dns:dns-tcp" -j KUBE-NOOP
-A KUBE-SEP-DERWY4MTP2RTOITK -s 10.64.1.9/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-DERWY4MTP2RTOITK -p tcp -m tcp -j DNAT ...
-A KUBE-SEP-HIAAPJKAI3E74A5J -m comment --comment "kube-system/kube-dns:dns" -j KUBE-NOOP
-A KUBE-SEP-HIAAPJKAI3E74A5J -s 10.64.1.9/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-HIAAPJKAI3E74A5J -p udp -m udp -j DNAT ...
-A KUBE-SEP-J2STD6VR5QCDF6IS -m comment --comment "kube-system/kube-dns:dns-tcp" -j KUBE-NOOP
-A KUBE-SEP-J2STD6VR5QCDF6IS -s 10.64.2.3/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-J2STD6VR5QCDF6IS -p tcp -m tcp -j DNAT ...
-A KUBE-SEP-SIPEKJQGVNBU5CE6 -m comment --comment "default/hostnames" -j KUBE-NOOP
-A KUBE-SEP-SIPEKJQGVNBU5CE6 -s 10.64.3.9/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-SIPEKJQGVNBU5CE6 -p tcp -m tcp -j DNAT ...
```

Reply:
> In an earlier comment I suggested maybe comment-only rules. That could help maybe?

That would mean more total rules, which is definitely a bad thing... sometimes... ("More rules" means both slightly slower save/restore and probably also slower "runtime"; I'm pretty sure the `-j KUBE-NOOP` won't get optimized out and the kernel will actually waste time figuring out it has to do nothing each time it hits one.) But then, having only one copy of each comment instead of multiple would probably make save/restore slightly faster... we'd have to do perf testing to figure out if it ended up being a net win or loss.

I had also been thinking about changing it to only comment on one rule in each chain. eg:

```
-A KUBE-SEP-5QBLSAUHS7VAMKHU -s 10.64.3.3/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-5QBLSAUHS7VAMKHU -m comment --comment "kube-system/metrics-server" -p tcp -m comment -m tcp -j DNAT ...
-A KUBE-SEP-5QBLSAUHS7VAMKHU -s 10.64.1.9/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-6IIMKUPVCRULMU5Y -m comment --comment "kube-system/kube-dns:metrics" -p tcp -m tcp -j DNAT ...
-A KUBE-SEP-6QKF7KDE3IMNRERG -s 10.64.2.7/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-6QKF7KDE3IMNRERG -m comment --comment "default/hostnames" -p tcp -m tcp -j DNAT ...
```

Both of these approaches are much less `iptables-save | grep`-friendly than the current system...